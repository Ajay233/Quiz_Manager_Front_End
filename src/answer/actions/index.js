import { getUsingParams, post, put, del } from '../../axiosRequests/requests'
import { setNotification } from '../../notifications/actions'
import { hideModal } from '../../modal/actions'
import { sessionExpired } from '../../utils/session'
import history from '../../history'

// create
export const addAnswer = (data, jwt) => {
  return (dispatch) => {
    return post("answer/create", [data], jwt).then((response) => {
      dispatch({
        type: "ADD_ANSWER",
        payload: response.data[0]
      })
      history.push("/editQuestion");
      dispatch(setNotification("Answer created", "success", true));
    }).catch((error) => {
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else {
        console.log(error.response);
        dispatch(setNotification("Error - unable to create answer", "error", true));
      }
    });
  }
}

// read
export const getAnswers = (param, jwt) => {
  return (dispatch) => {
    return getUsingParams("answer/findByQuestionId", param, jwt).then((response) => {
      dispatch({
        type: "SET_ANSWERS",
        payload: response.data
      });
    }).catch((error) => {
      dispatch(setNotification("Error retrieving answers", "error", true));
    });
  }
}

// update (there was a call to setCurrentAnswer but have removed this as it's not needed)
export const updateAnswer = (body, jwt) => {
  return (dispatch) => {
    return put("answer/update", [body], jwt).then((response) => {
      dispatch({
        type: "UPDATE_ANSWER",
        payload: response.data[0]
      })
      history.push("/editQuestion");
      dispatch(setNotification("Answer updated", "success", true));
    }).catch((error) => {
      console.log(error.response);
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else {
        dispatch(setNotification("Error - unable to update answer", "error", true));
      }
    });
  }
}

// delete
export const deleteAnswer = (config, jwt) => {
  return (dispatch) => {
    return del("answer/delete", config, jwt).then((response) => {
      dispatch(hideModal())
      dispatch({
        type: "DELETE_ANSWER",
        payload: config.data[0]
      })
      dispatch(setNotification("Answer deleted", "success", true));
    }).catch((error) => {
      console.log(error.response);
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else {
        dispatch(hideModal())
        dispatch(setNotification("Error - Unable to delete this answer", "error", true))
      }
    });
  }
}

export const setCurrentAnswer = (answer) => {
  return {
    type: "SET_CURRENT_ANSWER",
    payload: answer
  }
}
