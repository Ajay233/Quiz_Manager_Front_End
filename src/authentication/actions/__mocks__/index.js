export const setUser = jest.fn()

export const setVerficationProcess = jest.fn()

export const setRedirect = jest.fn()

export const logOut = jest.fn()
