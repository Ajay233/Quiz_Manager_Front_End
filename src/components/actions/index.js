export const setToTopButton = (show=false) => {
  return {
    type: "SHOW_TO_TOP_BUTTON",
    payload: show
  }
}

export const setUrl = (url="") => {
  return {
    type: "SET_URL",
    payload: url
  }
}

export const enableLightMode = () => {
  return {
    type: "SET_THEME",
    payload: false
  }
}

export const enableDarkMode = () => {
  return {
    type: "SET_THEME",
    payload: true
  }
}

export const setLoaderState = (show=false, message="", label="") => {
  return {
    type: "SET_LOADER_STATE",
    payload: {
      show: show,
      message: message,
      label: label
    }
  }
}
