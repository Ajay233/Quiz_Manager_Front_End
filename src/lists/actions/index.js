import { get } from '../../axiosRequests/requests'
import { sessionExpired } from '../../utils/session'

export const getCategories = (endpoint, jwt) => {
  return (dispatch) => {
    get(endpoint, jwt).then((response) => {
      dispatch({
        type: "SET_CATEGORIES",
        payload: response.data
      })
    }).catch((error) => {
      console.log(error.response)
      if(error.response.status === 403){
        sessionExpired(dispatch);
      }
    })
  }
}
