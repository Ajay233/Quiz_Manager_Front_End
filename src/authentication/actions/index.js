import { post } from '../../axiosRequests/requests'
import { setNotification } from '../../notifications/actions'
import history from '../../history'

export const setUser = (endpoint, loginDetails) => {

  return (dispatch) => {

    post(endpoint, loginDetails).then((response) => {
      dispatch({
        type: "SET_USER_LOGGED_IN",
        payload: {
          id: response.data.user.id,
          forename: response.data.user.forename,
          surname: response.data.user.surname,
          email: response.data.user.email,
          permission: response.data.user.permission,
          verified: response.data.user.verified,
          jwt: response.data.jwt,
          loggedIn: true,
          response: response
        }
      })
      history.push("/")
      dispatch(setNotification(`Welcome back ${response.data.user.forename}`, "loginSuccess", true))
    }).catch((error) => {
      if(error.response.data === "NOT VERIFIED"){
        const verifyMsg = "Your email has not yet been veirified.  You will need to verify your email before you can log in"
        dispatch(setNotification(verifyMsg, "warning", true))
      } else {
        dispatch({
          type: "SET_USER_LOGGED_IN",
          payload: {
            error: error.response
          }
        });
        dispatch(setNotification("The username or password you entered was incorrect", "error", true));
      }
    })
  }
}

export const logOutUser = () => {
  return {
    type: "LOG_OUT_USER",
    payload: {
      forename: "",
      surname: "",
      email: "",
      permission: "",
      verified: "",
      jwt: "",
      loggedIn: false
    }
  }
}

export const setVerficationProcess = (endpoint, verificationDetails) => {

  return (dispatch) => {

    post(endpoint, verificationDetails).then((response) => {
      dispatch({
        type: "SET_VERIFY_PROCESS_STATUS",
        payload: {
          completionStatus: "completed",
          token: "",
          error: {}
        }
      })
    }).catch((error) => {
      dispatch({
        type: "SET_VERIFY_PROCESS_STATUS",
        payload: {
          completionStatus: "not verified",
          token: verificationDetails.token,
          error: error.response
        }
      })
    })
  }
}

export const setRedirect = (status) => {
  return {
    type: "SET_REDIRECT",
    payload: {
      status: status
    }
  }
}

export const logOut = () => {
  return {
    type: "RESET_APP"
  }
}

// export const getUsers = (endpoint, header) => {
//   return {
//     type: "GET_USER",
//     payload: {
//       userId
//   }
// }
