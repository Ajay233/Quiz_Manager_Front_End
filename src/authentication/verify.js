import React from 'react';
import { connect } from 'react-redux';
import { Link, Redirect } from 'react-router-dom';
import Notification from '../notifications/notifications';

import { setVerficationProcess } from './actions'
import { setNotification } from '../notifications/actions'
import { post } from '../axiosRequests/requests'
import history from '../history'
import '../stylesheets/notifications.css'
import '../stylesheets/verify.css'


export class Verify extends React.Component {

  componentWillMount(){
    this.verifyUser();
  }

  verifyingMsg = () => {
    return <div>Verifying...</div>
  }

  verifiedMsg = () => {
    const { setNotification } = this.props
    const successMsg = "Your email has now been verified.  Please log in below to continue."
    setNotification(successMsg, "success", true);
    history.push("/login")
    // return <Redirect to="/login" />
  }

  notFoundMsg = () => {
      return (
        <div>
        <div id="verifyTitle">
          <div id="titleImg" >
            <img src={require("../public/icons/warning.png")} alt="" className="verifyTitleImg"/>
          </div>
          <div id="verifyTitleTxt" >Oops 404</div>
        </div>
          <p>Something went wrong, and the token you provided was not found.</p>
          <p>Possible causes can include:</p>
          <ul>
            <li>- You may have already verified your email address.  To check please click <Link to="/login">here</Link> to log in</li>
            <br/>
            <li>- If you copied and pasted the link into the browser address bar, you will need to check you did not miss any characters</li>
            <br/>
            <li>- You have accessed this page in error, in which case you will need to return to the <Link to="/">home</Link> page</li>
          </ul>
        </div>
      );
  }

  tokenExpiredMsg = () => {
      return (
        <div>
          <div id="verifyTitle">
            <div id="titleImg" >
              <img src={require("../public/icons/warning.png")} alt="" className="verifyTitleImg"/>
            </div>
            <div id="verifyTitleTxt" >Token Expired</div>
          </div>
          <p>Your token has expired as it was issued more than 24 hours ago.</p>
          <p>Please click resend to resend a new token</p>
          <button id="resendButton" className="resend" onClick={this.handleResend}>Resend</button>
        </div>
      );
  }

  handleResend = () => {
    const successMsg = "Token resent to the email provided during signup, please check your email and use the link provided to complete the sign up process";
    console.log(this.props.verificationProcess.token);
    const data = {'userId': 0, 'token': this.props.verificationProcess.token}
    post("auth/resendToken", data).then((response) => {
      this.props.setNotification(successMsg, "success", true);
    }).catch((error) => {
      console.log(error.response)
      this.props.setNotification(error.response.data, "error", true);
    })
  }

  getToken = () => {
    // let url = window.location.href;
    let url = history.location.search;
    const regex = new RegExp(/(?<=token=)(s?)(.*)/gm);
    let token = url.match(regex).toString();
    return token;
  }

  verifyUser = (event) => {
    console.log("VERIFYING.....")
    let token = this.getToken();
    const verificationDetails = {'userId': 0, 'token': token}
    this.props.setVerficationProcess('auth/verify', verificationDetails);
  }

  renderLogo = () => {
    const { data } = this.props.verificationProcess.error;
    if(data === "TOKEN_EXPIRED"){
      return <img id="contentImg" src={require("../public/icons/expired.png")} alt="" />
    } else if(data === "TOKEN_UNMATCHED") {
      return <img id="contentImg" src={require("../public/icons/notFound.png")} alt=""/>
    } else {
      return null;
    }
  }

  renderMsg = () => {
    const { data } = this.props.verificationProcess.error;
    if(data === "TOKEN_EXPIRED"){
      return this.tokenExpiredMsg();
    } else if(data === "TOKEN_UNMATCHED") {
      return this.notFoundMsg();
    } else {
      return null;
    }
  }

  render(){
    return(
      <div id="verifyContainer">
        {console.log(this.props.verificationProcess)}
        {this.props.verificationProcess.completionStatus === "" ? this.verifyingMsg() : null}
        {this.props.verificationProcess.completionStatus === "completed" ? this.verifiedMsg() : null}
        <Notification />
        <div id="verifyContent">
          <div id="verifyLogo">
            {this.renderLogo()}
          </div>
          <div id="verifyMessage">
            {this.renderMsg()}
          </div>
        </div>
      </div>
    );
  }
}

export const mapStateToProps = (state) => {
  return {
    userData: state.userData,
    verificationProcess: state.verificationProcess
  }
}

export default connect(mapStateToProps, {setVerficationProcess, setNotification})(Verify);
