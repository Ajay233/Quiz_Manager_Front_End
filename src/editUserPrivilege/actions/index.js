import { getUsingParams } from '../../axiosRequests/requests'
import { setNotification } from '../../notifications/actions'
import { sessionExpired } from '../../utils/session'

export const setUserResults = (param, jwt) => {
  return (dispatch) => {
    return getUsingParams("users/findByEmail", param, jwt).then((response) => {
      dispatch({
        type: "SET_USER_RESULTS",
        payload: response.data
      })
    }).catch((error) => {
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else {
        const msg = error.response.data;
        dispatch(setNotification(msg, "error", true));
        console.log(error.response)
      }
    });
  }
}

export const clearUserResults = () => {
  return {
    type: "CLEAR_USER_RESULTS",
    payload: {}
  }
}
