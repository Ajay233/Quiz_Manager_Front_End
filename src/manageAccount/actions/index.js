import { put } from '../../axiosRequests/requests'
import { setUserData } from '../../authentication/actions'
import { setNotification } from '../../notifications/actions'
import { sessionExpired } from '../../utils/session'

export const updateUser = (data, jwt) => {
  const emailChangeMsg = "Profile information updated - Your email was changed so your access has been restriced until your email is verified"
  const errorMsg = "Error updating your profile data, please check the details you provided and try again";
  const emailTaken = "An account with that email already exists"
  return (dispatch) => {
    return put('users/update', data, jwt).then((response) => {
      if(response.data.jwt === undefined){
        dispatch(setUserData(response.data, jwt))
        dispatch(setNotification("Profile information updated", "success", true))
      } else {
        dispatch(setUserData(response.data.user, response.data.jwt))
        dispatch(setNotification(emailChangeMsg, "success", true))
      }
    }).catch((error) => {
      console.log(error)
      console.log(error.response)
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else if(error.response.data === emailTaken){
        dispatch(setNotification(emailTaken, "error", true))
      } else {
        dispatch(setNotification(errorMsg, "error", true))
      }
    });
  }
}
