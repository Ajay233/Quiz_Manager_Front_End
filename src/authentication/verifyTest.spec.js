import React from 'react'
import { Provider } from 'react-redux'
import { Router } from 'react-router-dom'
import Verify from './verify'
import { mapStateToProps } from './verify'
import { shallow, mount } from 'enzyme'
import renderer from 'react-test-renderer'
import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import mockAxios from 'jest-mock-axios'
import history from '../history'
import { setVerficationProcess } from './actions'
import { setNotification } from '../notifications/actions'
import { setLoaderState } from '../components/actions'

// jest.mock('./actions')
// jest.mock('../notifications/actions')

const middlewares = [thunk]
const mockStore = configureStore(middlewares)

describe("mapStateToProps", () => {
  it("should map state to props", () => {
    const appState = {
      userData: { id: 1, jwt: "jwt" },
      verificationProcess: { completionStatus: "test", token: "test", error: {} }
    }
    const componentState = mapStateToProps(appState)
    expect(componentState).toEqual(appState)
  })
})

describe("Verify component", () => {

  afterEach(() => {
    mockAxios.reset()
  })

  // it("should call verified message and the methods within that", () => {
  //
  //   ------ONLY WORKS IF USING COMPONENTWILLMOUNT BUT THIS CAN NO LONGER BE USED-----
  //
  //   const userData = { id: 1, jwt: "jwt" }
  //   const verificationProcess = { completionStatus: "test", token: "test", error: {} }
  //   const notificationData = { message: "", type: "", show: false, timed: true }
  //
  //   const store = mockStore({
  //     userData: { id: 1, jwt: "jwt" },
  //     verificationProcess: { completionStatus: "completed", token: "", error: {} },
  //     notificationData: { message: "", type: "", show: false, timed: true }
  //   })
  //
  //   history.push("/verify?token=test");
  //
  //   const wrapper = mount(
  //     <Provider store={store}>
  //       <Router history={history}>
  //         <Verify
  //           userData={userData}
  //           verificationProcess={verificationProcess}
  //           notificationData={notificationData}
  //           setVerficationProcess={setVerficationProcess}
  //           setNotification={setNotification}
  //         />
  //       </Router>
  //     </Provider>
  //   );
  //
  //   // console.log(wrapper.debug())
  //   const successMsg = "Your email has now been verified.  Please log in below to continue."
  //   expect(history.location.pathname).toEqual("/login")
  //   // expect(setNotification).toHaveBeenCalledTimes(1)
  //   // expect(setNotification).toHaveBeenCalledWith(successMsg, "success", true)
  // })

  it("should render the token expired message when error data is set to 'TOKEN_EXPIRED'", () => {

    const userData = { id: 1, jwt: "jwt" }
    const verificationProcess = { completionStatus: "test", token: "test", error: {} }
    const notificationData = { message: "", type: "", show: false, timed: true }

    const store = mockStore({
      userData: { id: 1, jwt: "jwt" },
      verificationProcess: { completionStatus: "test", token: "test", error: { data: "TOKEN_EXPIRED" } },
      notificationData: { message: "", type: "", show: false, timed: true },
      globals: { loaderState: { show: false } }
    })

    history.push("/verify?token=test");

    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <Verify
            userData={userData}
            verificationProcess={verificationProcess}
            notificationData={notificationData}
            setVerficationProcess={setVerficationProcess}
            setNotification={setNotification}
            setLoaderState={setLoaderState}
          />
        </Router>
      </Provider>
    );

    mockAxios.reset()
    wrapper.find('button').simulate('click')
    expect(mockAxios.post).toHaveBeenCalledTimes(1)
    // This would need to be changed back after deployment if working locally, so this will need to be reworked.
    // This network request may also need to be moved into an action creator.  If so that will change how this test is set up.
    // expect(mockAxios.post).toHaveBeenCalledWith("http://localhost:8080/auth/resendToken", {'userId': 0, 'token': "test" } )
  })

  it("should render the token expired message when error data is set to 'TOKEN_EXPIRED'", () => {

    const userData = { id: 1, jwt: "jwt" }
    const verificationProcess = { completionStatus: "test", token: "test", error: {} }
    const notificationData = { message: "", type: "", show: false, timed: true }

    const store = mockStore({
      userData: { id: 1, jwt: "jwt" },
      verificationProcess: { completionStatus: "test", token: "test", error: { data: "TOKEN_EXPIRED" } },
      notificationData: { message: "", type: "", show: false, timed: true },
      globals: { loaderState: { show: false } }
    })

    history.push("/verify?token=test");

    const component = renderer.create(
      <Provider store={store}>
        <Router history={history}>
          <Verify
            userData={userData}
            verificationProcess={verificationProcess}
            notificationData={notificationData}
            setVerficationProcess={setVerficationProcess}
            setNotification={setNotification}
            setLoaderState={setLoaderState}
          />
        </Router>
      </Provider>
    );

    expect(component).toMatchSnapshot()
  })

  it("should render the token expired message when error data is set to 'TOKEN_UNMATCHED'", () => {

    const userData = { id: 1, jwt: "jwt" }
    const verificationProcess = { completionStatus: "test", token: "test", error: {} }
    const notificationData = { message: "", type: "", show: false, timed: true }

    const store = mockStore({
      userData: { id: 1, jwt: "jwt" },
      verificationProcess: { completionStatus: "test", token: "test", error: { data: "TOKEN_UNMATCHED" } },
      notificationData: { message: "", type: "", show: false, timed: true },
      globals: { loaderState: { show: false } }
    })

    history.push("/verify?token=test");

    const component = renderer.create(
      <Provider store={store}>
        <Router history={history}>
          <Verify
            userData={userData}
            verificationProcess={verificationProcess}
            notificationData={notificationData}
            setVerficationProcess={setVerficationProcess}
            setNotification={setNotification}
            setLoaderState={setLoaderState}
          />
        </Router>
      </Provider>
    );

    expect(component).toMatchSnapshot()
  })

  it("should render the token expired message when error data is set to 'TOKEN_UNMATCHED'", () => {

    const userData = { id: 1, jwt: "jwt" }
    const verificationProcess = { completionStatus: "test", token: "test", error: {} }
    const notificationData = { message: "", type: "", show: false, timed: true }

    const store = mockStore({
      userData: { id: 1, jwt: "jwt" },
      verificationProcess: { completionStatus: "test", token: "test", error: {} },
      notificationData: { message: "", type: "", show: false, timed: true },
      globals: { loaderState: { show: false } }
    })

    history.push("/verify?token=test");

    const component = renderer.create(
      <Provider store={store}>
        <Router history={history}>
          <Verify
            userData={userData}
            verificationProcess={verificationProcess}
            notificationData={notificationData}
            setVerficationProcess={setVerficationProcess}
            setNotification={setNotification}
            setLoaderState={setLoaderState}
          />
        </Router>
      </Provider>
    );

    expect(component).toMatchSnapshot()
  })
})
