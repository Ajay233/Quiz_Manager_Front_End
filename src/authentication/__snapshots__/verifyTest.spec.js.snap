// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Verify component should render the token expired message when error data is set to 'TOKEN_EXPIRED' 1`] = `
<div
  id="verifyContainer"
>
  <div
    className="spacing"
    id="notification"
  />
  <div
    id="verifyContent"
  >
    <div
      id="verifyLogo"
    >
      <img
        alt=""
        id="contentImg"
        src="expired.png"
      />
    </div>
    <div
      id="verifyMessage"
    >
      <div>
        <div
          id="verifyTitle"
        >
          <div
            id="titleImg"
          >
            <img
              alt=""
              className="verifyTitleImg"
              src="warning.png"
            />
          </div>
          <div
            id="verifyTitleTxt"
          >
            Token Expired
          </div>
        </div>
        <p>
          Your token has expired as it was issued more than 24 hours ago.
        </p>
        <p>
          Please click resend to resend a new token
        </p>
        <button
          className="resend"
          id="resendButton"
          onClick={[Function]}
        >
          Resend
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Verify component should render the token expired message when error data is set to 'TOKEN_UNMATCHED' 1`] = `
<div
  id="verifyContainer"
>
  <div
    className="spacing"
    id="notification"
  />
  <div
    id="verifyContent"
  >
    <div
      id="verifyLogo"
    >
      <img
        alt=""
        id="contentImg"
        src="notFound.png"
      />
    </div>
    <div
      id="verifyMessage"
    >
      <div>
        <div
          id="verifyTitle"
        >
          <div
            id="titleImg"
          >
            <img
              alt=""
              className="verifyTitleImg"
              src="warning.png"
            />
          </div>
          <div
            id="verifyTitleTxt"
          >
            Oops 404
          </div>
        </div>
        <p>
          Something went wrong, and the token you provided was not found.
        </p>
        <p>
          Possible causes can include:
        </p>
        <ul>
          <li>
            - You may have already verified your email address.  To check please click 
            <a
              href="/login"
              onClick={[Function]}
            >
              here
            </a>
             to log in
          </li>
          <br />
          <li>
            - If you copied and pasted the link into the browser address bar, you will need to check you did not miss any characters
          </li>
          <br />
          <li>
            - You have accessed this page in error, in which case you will need to return to the 
            <a
              href="/"
              onClick={[Function]}
            >
              home
            </a>
             page
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;

exports[`Verify component should render the token expired message when error data is set to 'TOKEN_UNMATCHED' 2`] = `
<div
  id="verifyContainer"
>
  <div
    className="spacing"
    id="notification"
  />
  <div
    id="verifyContent"
  >
    <div
      id="verifyLogo"
    />
    <div
      id="verifyMessage"
    />
  </div>
</div>
`;
